mlflow/pyfunc/utils/data_validation.py

```diff
@@ -0,0 +1,122 @@
+import inspect
+import warnings
+from functools import lru_cache, wraps
+from typing import Any, Optional
+
+from mlflow.models.signature import (
+    _extract_type_hints,
+    _is_context_in_predict_function_signature,
+    _TypeHints,
+)
+from mlflow.types.type_hints import (
+    _convert_data_to_type_hint,
+    _infer_schema_from_type_hint,
+    _signature_cannot_be_inferred_from_type_hint,
+    _validate_example_against_type_hint,
+)
+from mlflow.utils.annotations import filter_user_warnings_once
+
+
+def _wrap_predict_with_pyfunc(func, type_hint):
+    if type_hint is not None:
+        model_input_index = _model_input_index_in_function_signature(func)
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            args, kwargs = _validate_model_input(args, kwargs, model_input_index, type_hint)
+            return func(*args, **kwargs)
+    else:
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            return func(*args, **kwargs)
+
+    wrapper._is_pyfunc = True
+    return wrapper
+
+
+def pyfunc(func):
+    """
+    A decorator that forces data validation against type hint of the input data
+    in the wrapped method. It is no-op if the type hint is not supported by MLflow.
+
+    .. note::
+        The function that applies this decorator must be a valid `predict` function
+        of `mlflow.pyfunc.PythonModel`, or a callable that takes a single input.
```

Comment:
q: Won't we validate input regardless of existence of @pyfunc annotation for `PythonModel`?

Reply:
Yes